File: projects/CSharp_ColledgeEdition-master/Time/TimeStruct.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int seconds;
} Time;

//инициализатор 1
Time createTime(int hours, int minutes, int seconds) {
    Time newTime;
    newTime.seconds = hours * 3600 + minutes * 60 + seconds;
    return newTime;
}

//инициализатор 2
Time fromDateTime(int hours, int minutes, int seconds) {
    return createTime(hours, minutes, seconds);
}

//инициализатор 3
Time fromTotalSeconds(int totalSeconds) {
    return createTime(0, 0, totalSeconds);
}

//инициализатор 4
Time fromTimeString(char* timeString) {
    int hours, minutes, seconds;
    if (sscanf(timeString, "%d:%d:%d", &hours, &minutes, &seconds) == 3) {
        return createTime(hours, minutes, seconds);
    }
    else {
        printf("Invalid time string format\n");
        exit(EXIT_FAILURE);
    }
}

//Разница секунд
int differenceInSeconds(Time time1, Time time2) {
    return abs(time1.seconds - time2.seconds);
}

//Добавление секунд
Time addSeconds(Time time, int secondsToAdd) {
    time.seconds += secondsToAdd;
    return time;
}

//метод вычитания секунд
Time subtractSeconds(Time time, int secondsToSubtract) {
    time.seconds -= secondsToSubtract;
    if (time.seconds < 0) {
        time.seconds = 0;
    }
    return time;
}

//метод проверки равенства
int equals(Time time1, Time time2) {
    return time1.seconds == time2.seconds;
}

//конвертация в секунды
int toSeconds(Time time) {
    return time.seconds;
}
//конвертация в минуты
int toMinutes(Time time) {
    return (int)(time.seconds / 60.0 + 0.5);
}

//Вывод времени
void printTime(Time time) {
    printf("%02d:%02d:%02d\n", time.seconds / 3600, (time.seconds % 3600) / 60, time.seconds % 60);
}



int main() {
    Time time1 = createTime(1, 30, 0);
    Time time2 = createTime(0, 45, 0);

    Time result1 = addSeconds(time1, toSeconds(time2));
    Time result2 = subtractSeconds(time1, toSeconds(time2));

    printTime(result1);
    printTime(result2);

    Time time3 = fromDateTime(12, 34, 56);
    Time time4 = fromTotalSeconds(5000);
    Time time5 = fromTimeString("02:15:30");

    printTime(time3);
    printTime(time4);
    printTime(time5);

    return 0;
}

File: projects/CSharp_ColledgeEdition-master/Time/TimeStruct.cs
namespace Time;

public struct Time
{
    private int Seconds;

    public Time(int hours, int minutes, int seconds)
    {
        this.Seconds = hours * 3600 + minutes * 60 + seconds;
        CheckSecondsBelowZero();
    }

    public Time(int totalSeconds) : this()
    {
        this.Seconds = totalSeconds;
        CheckSecondsBelowZero();
    }

    public static Time FromDateTime(DateTime time) 
        => new Time(time.Hour, time.Minute, time.Second);

    public static Time FromTotalSeconds(int totalSeconds)
    {
        return new Time(totalSeconds);
    }

    public static Time FromTimeString(string timeString)
    {
        string[] parts = timeString.Split(':');
        if (parts.Length == 3 && int.TryParse(parts[0], out int hours) && int.TryParse(parts[1], out int minutes) && int.TryParse(parts[2], out int secs))
        {
            return new Time(hours, minutes, secs);
        }
        else
        {
            throw new ArgumentException("Invalid time string format");
        }
    }

    #region methods
    private void CheckSecondsBelowZero() 
        => this.Seconds = this.Seconds < 0 ? 0 : this.Seconds;

    public static int DifferenceInSeconds(Time time1, Time time2)
        => Math.Abs(time1.Seconds - time2.Seconds);

    public static Time AddSeconds(Time time, int secondsToAdd)
    {
        time.Seconds += secondsToAdd;
        return time;
    }

    public static Time SubtractSeconds(Time time, int secondsToSubtract)
    {
        time.Seconds -= secondsToSubtract;
        if (time.Seconds < 0)
        {
            time.Seconds = 0;
        }
        return time;
    }

    public static bool Equals(Time time1, Time time2)
        => time1.Seconds == time2.Seconds;

    public static int ToSeconds(Time time)
        => time.Seconds;

    public static int ToMinutes(Time time)
        => (int)(time.Seconds / 60.0 + 0.5);

    public static void PrintTime(Time time)
    {
        Console.WriteLine($"{time.Seconds / 3600:D2}:{(time.Seconds % 3600) / 60:D2}:{time.Seconds % 60:D2}");
    }
    #endregion
}
/* также тестовый main
public class Program
{
    public static void Main()
    {
        Time time1 = new Time(1, 30, 0);
        Time time2 = new Time(0, 45, 0);

        Time result1 = Time.AddSeconds(time1, Time.ToSeconds(time2));
        Time result2 = Time.SubtractSeconds(time1, Time.ToSeconds(time2));

        Time.PrintTime(result1);
        Time.PrintTime(result2);

        Time time3 = Time.FromDateTime(DateTime.Now);
        Time time4 = Time.FromTotalSeconds(5000);
        Time time5 = Time.FromTimeString("12:34:56");

        Time.PrintTime(time3);
        Time.PrintTime(time4);
        Time.PrintTime(time5);
    }
}
*/